src/Ace/Halogen/Component.purs

```diff
@@ -1,75 +1,185 @@
+import Control.Monad.Eff.Ref (Ref(), REF(), readRef, writeRef, modifyRef)
+import DOM (DOM())
+import DOM.HTML.Types (HTMLElement())
+import Data.Date (nowEpochMilliseconds, Now())
+import Data.Foldable (traverse_)
+import Data.Maybe (Maybe(..), maybe)
+import Data.StrMap (StrMap())
+import Data.StrMap as Sm
+import Data.Time (Milliseconds(..))
+import Halogen
 import Halogen.HTML.Indexed as H
 import Halogen.HTML.Properties.Indexed as P
 
-import Ace.Types (ACE(), Editor())
-import Ace.Editor as Editor
-import Ace.Halogen.Component.Query
-import Ace.Halogen.Component.State
+----------------------------------------------------------------------
+-- Foreign knots and effectful functions
+----------------------------------------------------------------------
+foreign import completeFns :: forall e. Ref (StrMap (CompleteFn e))
+foreign import initialized :: Ref Boolean
+foreign import focused :: Ref String
+foreign import dataset
+  :: forall e. HTMLElement -> Eff (dom :: DOM |e) (StrMap String)
+
+completeFnFocused :: forall e. Eff (AceEffects e) (CompleteFn e)
+completeFnFocused = do
+  focusedKey <- readRef focused
+  mFns <- readRef completeFns
+  maybe (pure emptyCompleteFn) pure $ Sm.lookup focusedKey mFns
+  where
+  emptyCompleteFn :: CompleteFn e
+  emptyCompleteFn _ _ _ _ = pure []
+
+setAutocompleteResume
+  :: forall e. Maybe Autocomplete -> Editor -> Eff (AceEffects e) Unit
+setAutocompleteResume Nothing editor = do
+  Editor.setEnableBasicAutocompletion false editor
+  Editor.setEnableLiveAutocompletion false editor
+setAutocompleteResume (Just Basic) editor = do
+  Editor.setEnableLiveAutocompletion false editor
+  Editor.setEnableBasicAutocompletion true editor
+setAutocompleteResume (Just Live) editor = do
+  Editor.setEnableLiveAutocompletion true editor
+  Editor.setEnableBasicAutocompletion true editor
+
+enableAutocomplete :: forall e. Eff (AceEffects e) Unit
+enableAutocomplete = do
+  languageToolsInitialized <- readRef initialized
+  when (not languageToolsInitialized) do
+    completer <- Completer.mkCompleter globalCompleteFn
+    tools <- LanguageTools.languageTools
+    LanguageTools.addCompleter completer tools
+    writeRef initialized true
+  where
+  globalCompleteFn editor session position prefix cb = do
+    fn <- completeFnFocused
+    runAff (const $ cb Nothing) (cb <<< Just)
+      $ fn editor session position prefix
+
+genKey :: forall e. Eff (now :: Now, random :: RANDOM|e) String
+genKey = do
+  rn1 <- random
+  rn2 <- random
+  (Milliseconds time) <- nowEpochMilliseconds
+  pure $ show rn1 <> show time <> show rn2
+
+----------------------------------------------------------------------
+-- Types, State and other common halogen things
+----------------------------------------------------------------------
+data Autocomplete = Live | Basic
+
+type AceEffects e =
+  ( random :: RANDOM
+  , now :: Now
+  , ref :: REF
+  , ace :: ACE
+  , avar :: AVAR
+  , dom :: DOM
+  | e)
+
+data AceQuery a
```

Comment:
Can we restore/update the comment here that describes what the constructors are for?